src/vma/lwip/tcp_out.c

```diff
@@ -848,6 +849,125 @@ tcp_send_empty_ack(struct tcp_pcb *pcb)
+	u32_t lentosend = (wnd - (seg->seqno - pcb->lastack));
+	u16_t oversize = 0;
+	u8_t  optlen = 0;
+	u8_t  apiflags = 0;
+	u8_t  optflags = 0;
+
+	if (seg->p->ref>1) {
+		return;
+	}
+
+	#if LWIP_TCP_TIMESTAMPS
+	if ((pcb->flags & TF_TIMESTAMP)) {
+		optflags = TF_SEG_OPTS_TS;
+	}
+	#endif /* LWIP_TCP_TIMESTAMPS */
+
+	optlen = LWIP_TCP_OPT_LENGTH( optflags );
+
+	if(seg->p->len > lentosend) {//first buffer is too big, split it
+
+		u32_t lentoqueue = seg->p->len - TCP_HDR - lentosend;
+
+		if (NULL == (p = tcp_pbuf_prealloc(lentoqueue + optlen, lentoqueue + optlen, &oversize, pcb, apiflags, 0))) {
+			LWIP_DEBUGF(TCP_OUTPUT_DEBUG | 2, ("tcp_output : could not allocate memory for pbuf copy size %"U16_F"\n", (lentoqueue + optlen)));
+			return;
+		}
+
+		/* Copy the data from the original buffer */
+		TCP_DATA_COPY2((char *)p->payload + optlen, (u8_t *)seg->dataptr + lentosend, lentoqueue , &chksum, &chksum_swapped);
+
+		/*Update new buffer*/
+		p->tot_len = seg->p->tot_len - lentosend - TCP_HDR;
+		p->next = seg->p->next;
+
+		/*Update original buffer*/
+		seg->p->next = NULL;
+		seg->p->len = seg->p->len - lentoqueue;
+		seg->p->tot_len = seg->p->len;
+
+		/* Allocate memory for tcp_seg, and fill in fields. */
+		if (NULL == (newseg = tcp_create_segment(pcb, p, 0,  seg->seqno + lentosend, optflags))) {
+			LWIP_DEBUGF(TCP_OUTPUT_DEBUG | 2, ("tcp_output : could not allocate memory for segment\n"));
+			tcp_tx_pbuf_free(pcb, p);
+			return;
+		}
+
+		/*New segment update*/
+		newseg->next = seg->next;
+		newseg->flags = seg->flags;
+
+		/* Original segment update */
+		seg->next = newseg;
+		seg->len = seg->p->len - TCP_HDR;
+
+		if (pcb->last_unsent == seg) {
+		  pcb->last_unsent = newseg;
+		  pcb->unsent_oversize = oversize;
+		}
+
+		/* Set the PSH flag in the last segment that we enqueued. */
+		TCPH_SET_FLAG(newseg->tcphdr, TCP_PSH);
+
+		/* Update number of buffer to be send */
+		pcb->snd_queuelen++;
+	}
+	else {
+		struct pbuf *pnewhead = seg->p->next;
+		struct pbuf *pnewtail = seg->p;
+		struct pbuf *ptmp = seg->p;
+		u32_t headchainlen = seg->p->len;
+
+		while ((headchainlen + pnewhead->len - TCP_HDR)<= lentosend) {
+			headchainlen += pnewhead->len;
+			pnewtail = pnewhead;
+			pnewhead = pnewhead->next;
+		}
+
+		/* Allocate memory for tcp_seg, and fill in fields. */
+		if (NULL == (newseg = tcp_create_segment(pcb, pnewhead, 0,  seg->seqno + headchainlen - TCP_HDR, optflags))) {
+			LWIP_DEBUGF(TCP_OUTPUT_DEBUG | 2, ("tcp_output : could not allocate memory for segment\n"));
+			tcp_tx_pbuf_free(pcb, p);
+			return;
+		}
+
+		/*Update new tail*/
+		pnewtail->next = NULL;
+
+		/*New segment update*/
+		newseg->next = seg->next;
+		newseg->flags = seg->flags;
+
+		/* Original segment update */
+		seg->next = newseg;
+		seg->len = headchainlen - TCP_HDR;
+
+		/*Update original buffers*/
+		while(ptmp) {
+			ptmp->tot_len = headchainlen;
+			headchainlen -= ptmp->len;
+			ptmp = ptmp->next;
+		}
+	}
+
+    {
```

Comment:
formatting: bracket indentation is based on 4 space, while rest of file is 1xTAB